context.go

```diff
@@ -0,0 +1,141 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build (linux || darwin) && (amd64 || arm64)
+
+package waf
+
+import (
+	"runtime"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+// Context is a WAF execution context.
+// It's a wrapper over the void * the WAF returns with ddwaf_context_init()
+// We store in there:
+// * A mutex to order calls to ddwaf_run()
+// * The handle it was created with
+// * The allocator to keep references to the go waf objects we sent
+// * A number of metrics
+type Context struct {
+	handle    *Handle
+	cContext  wafContext
+	mutex     sync.Mutex
+	allocator allocator
+
+	// Stats
+	// Cumulated internal WAF run time - in nanoseconds - for this context.
+	totalRuntimeNs atomic.Uint64
+	// Cumulated overall run time - in nanoseconds - for this context.
+	totalOverallRuntimeNs atomic.Uint64
+	// Cumulated timeout count for this context.
+	timeoutCount atomic.Uint64
+}
+
+// Run encode the input data to send it to the WAF and returns the return values of ddwaf_run()
+// The parameters are the following:
+// * addressesToData: a map of key WAF addresses to the user input. The user input can be any go type.
+// * timeout: a timeout value to run the waf. If the threshold is exceeded, ddwaf_run will return without completing everything
+// The return values are the following:
+// * matches: A json string directly sent by the WAF containing all the data necessary for the backend
+// * actions: A array of string referencing which _blocking_ rules have been triggered
+// * err: a RunError error from the error codes the WAF returns
+func (context *Context) Run(addressesToData map[string]any, timeout time.Duration) (matches []byte, actions []string, err error) {
+	if len(addressesToData) == 0 {
+		return
+	}
+
+	now := time.Now()
+	defer func() {
+		dt := time.Since(now)
+		context.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
+	}()
+
+	encoder := encoder{
+		stringMaxSize:    wafMaxStringLength,
+		containerMaxSize: wafMaxContainerSize,
+		objectMaxDepth:   wafMaxContainerDepth,
+	}
+	obj, err := encoder.Encode(addressesToData)
```

Comment:
After reviewing the purpose and usage of the allocator, I think this top-level `wafObject` is missing in the list of ref-counted Go values in `encoder.allocator`.